R/data.table.R

```diff
@@ -1890,37 +1890,45 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
       stop("length(rownames.value)==", length(rownames.value),
            " but should be nrow(x)==", nrow(x))
   }
-  if (!is.null(rownames)) {
-    # extract that column.
-    if (is.ff(x[[rownames]]))
-      rownames.value = x[[rownames]][] # bring the column into memory if ff object
-    else
-      rownames.value = x[[rownames]]
-    dm = dim(x) - 0:1
-    cn = names(x)[-rownames]
-  } else {
-    dm = dim(x)
-    cn = names(x)
-    rownames = ncol(x) + 1L # if no rownames to drop, set index to > ncol so that we can do [-rownames] rather than if (is.null(rownames)) {} else {} to drop the column in the rest of the code
-  }
+  # Create a new list X containing pointers to columns in x, so that any
+  # modifications (i.e. type coercion, dropping of rownames column) do 
+  # not change the input data.table. New copies in memory are only 
+  # created for columns that are modified (those coerced to a different
+  # type or class)
   X = x
-  if (any(dm == 0L))
-    return(array(NA, dim = dm, dimnames = list(rownames.value, cn)))
+  dm = dim(x)
+  cn = names(x)
+  class(X) = NULL
   p = length(cn)
   n = dm[1L]
-  collabs = as.list(cn)
```

Comment:
I could not find a reason for the line `collabs = as.list(cn)` - the only place `collabs` is now used is when given to `dimnames` after the matrix creation, in which it is unlisted `unlist(collabs, use.names=FALSE)`. I've now removed this, using simply `cn`, the names of the input data.table. All tests still pass.

Reply:
IIRC originally this was so `unlist(collabs)` would properly handle the nested column naming

Reply:
This is definitely no longer needed now we're using as.data.table to expand these columns and redetermine the column names and matrix dimensions.